EIP-010-wasi-command-runner.md

```diff
@@ -0,0 +1,68 @@
+
+## Solution
+
+The interface would take in a buffer to use as STDIN data and return two buffers out, STDOUT and STDERR. As WASI command modules are setup to work with pipes for STDIN input data and STDOUT and STDERR output data, setting this up in various Web Assembly runtimes is non-trivial.
+
+Environment variables and CLI args would also be passed as parameters.
+
+For efficiency reasons, ideally the solution would only compile the module once, but instantiate the module on every call to run it as it's unsafe to run WASI module multiple times.
+
+The persistent object holding the compiled module:
+```rust
+pub struct WASICommand {
+    engine: wasmtime::Engine,
+    module: wasmtime::Module,
+}
+```
+
+The run function:
+```rust
+    pub fn run<'a, T: ToBytes<'a>, U: FromBytesOwned, V: FromBytesOwned>(
+        &self,
+        envs: &[(String, String)],
+        args: &[String],
+        stdin: T,
+    ) -> Result<(U, V), Error> {
+        let bytes = stdin.to_bytes()?;
+        let stdin = ReadPipe::from(bytes.as_ref());
+        let stdout = WritePipe::new_in_memory();
+        let stderr = WritePipe::new_in_memory();
+        let wasi_ctx = wasmtime_wasi::WasiCtxBuilder::new()
+            .args(args)?
+            .envs(envs)?
+            .stdin(Box::new(stdin.clone()))
+            .stdout(Box::new(stdout.clone()))
+            .stderr(Box::new(stderr.clone()))
+            .build();
+
+        let mut store = wasmtime::Store::new(&self.engine, wasi_ctx);
+        let mut linker = wasmtime::Linker::new(&self.engine);
+        wasmtime_wasi::add_to_linker(&mut linker, |wasi| wasi)?;
+        let instance = linker.instantiate(&mut store, &self.module)?;
+        let function_name = "_start";
+        let f = instance
+            .get_func(&mut store, function_name)
+            .expect("function exists");
+        f.call(&mut store, &[], &mut []).unwrap();
+        drop(store);
+        let contents: Vec<u8> = stdout.try_into_inner().unwrap().into_inner();
+        let stderr_contents: Vec<u8> = stderr.try_into_inner().unwrap().into_inner();
+        let stdout_output = U::from_bytes_owned(&contents)?;
+        let stderr_output = V::from_bytes_owned(&stderr_contents)?;
+        Ok((stdout_output, stderr_output))
+    }
+}
+```
+
+https://github.com/extism/extism/tree/g4vi/wasi-command-runner
+
+## Considerations
+
+### Integration with existing Extism ecosystem manifest
+
+Initially this wouldn't support the manifest at all. Supporting the/a manifest would require further consideration due to it having configuration items not applicable to non-plugin modules.
```

Comment:
Great idea! Some kind of support for command modules with access to stdin/stdout/stderr would be a nice feature for Extism, but I think we could do a little more in terms of integrating it into the existing API. 

Something similar could be achieved by adding new function like `Plugin::call_command<Stdout: FromBytesOwned, Stderr: FromBytesOwned>(input: impl ToBytes) -> Result<(Stdout, Stderr), Error>` (or similar) that hooks up all the streams and executes the command plugin closer to how `Plugin::call` works, that way we would still be able to use timeouts/memory limits and things like that. 

We would also probably want to consider updating the C API so this would also be usable from the other SDKs.

Reply:
I took a closer look at integrating it with `Plugin`. A major issue to work around is handling differences in linking. `Plugin` currently integrates wasi in `build_new` with `wasmtime_wasi::add_to_linker`; the contents of `stdin` and `stdout` need to be specified before linking in wasi.

Any recommendations on how to proceed? I see the when the plugin function is `_start`, the store is reset and pre-linking on is redone, but wasi is only initialized/added to linker once in `build_new`.

Reply:
In `Plugin::reset_store` (https://github.com/extism/extism/blob/054a29e91d102ffa052c4d445ad4f2fef272d579/runtime/src/plugin.rs#L366) this call to `CurrentPlugin::new` https://github.com/extism/extism/blob/054a29e91d102ffa052c4d445ad4f2fef272d579/runtime/src/plugin.rs#L375 is actually rebuilding the WASI context each time too: https://github.com/extism/extism/blob/054a29e91d102ffa052c4d445ad4f2fef272d579/runtime/src/current_plugin.rs#L290

I think we could try adding an optional argument to `CurrentPlugin::new` to specify stdin and thread that through `Plugin::reset_store` - then from `Plugin::call_command` we could call `Plugin::reset_store` with that argument set. 

It would probably make sense to copy `Plugin::call` and switch out some of the code before and after the actual call to get access to avoid setting the Extism input and to get access to the WASI output.